src/mame/sinclair/specnext_divmmc.cpp

```diff
@@ -0,0 +1,136 @@
+// license:BSD-3-Clause
+// copyright-holders:Andrei I. Holub
+/**********************************************************************
+    Spectrum Next DivMMC
+**********************************************************************/
+
+#include "emu.h"
+#include "specnext_divmmc.h"
+
+
+specnext_divmmc_device::specnext_divmmc_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: device_t(mconfig, SPECNEXT_DIVMMC, tag, owner, clock)
+{
+}
+
+bool specnext_divmmc_device::conmem()
+{
+	return BIT(m_divmmc_reg, 7);
+}
+
+bool specnext_divmmc_device::mapram()
+{
+	return BIT(m_divmmc_reg, 6);
+}
+
+bool specnext_divmmc_device::page0()
+{
+	return m_cpu_a_15_13 == 0;
+}
+
+bool specnext_divmmc_device::page1()
+{
+	return m_cpu_a_15_13 == 1;
+}
+
+bool specnext_divmmc_device::rom_en()
+{
+	return page0() && (conmem() || automap()) && !mapram();
+}
+
+bool specnext_divmmc_device::ram_en()
+{
+	return (page0() && (conmem() || automap()) && mapram()) || (page1() && (conmem() || automap()));
+}
```

Comment:
The other thing about these is that it’s not immediately clear whether the thing with side effects will happen, because it’s deep in a Boolean expression with short-circuit operators.

```
bool specnext_divmmc_device::rom_en()
{
	return page0() && (conmem() || automap()) && !mapram();
}

bool specnext_divmmc_device::ram_en()
{
	return (page0() && (conmem() || automap()) && mapram()) || (page1() && (conmem() || automap()));
}
```

Reply:
Oh never mind, the functions no longer has side effects.  That’s easier to understand.